### PR TITLE
Add the Content caching feature

### PR DESCRIPTION
**Pull request by hsubox76**
# Feature: Content Caching
Design doc (API) (internal): https://docs.google.com/document/d/1kJTdoz-RO7mkzsjE1QDfJfD1YvSxJXuzmqTAtsS1hJ0/edit?resourcekey=0-cv2C2J8SLDn__yoetTPvBw&tab=t.0
Design doc (Python SDK) (internal): https://docs.google.com/document/d/11en7rPCuXIu-hfF71jUDWkfQctWeFEoq_5KgkinnxWg/edit?resourcekey=0-ntvB0-Y_GBQXOfkO-KDaXg&tab=t.0

## APIs added
- GoogleAICacheManager class (exported from new /server path) with create, update, get, list, delete methods.
- GoogleGenerativeAI.getGenerativeModelFromContextCache() added

## Infrastructure changed
- `/files` subpath changed to `/server`. GoogleAIFileManager and GoogleAICacheManager will be exported from here as they are only meant to be used in a server environment.
- All types moved to a single folder.
- Generate common/api-review md files for easier review of API mistakes/gaps
- Refactor `makeRequest`-related functions to reduce code duplication

### CODE REVIEW

File: packages/main/src/requests/request.ts [hunk 1]
```diff
@@ -130,89 +130,86 @@ export async function constructRequest(
...
-        response.status,
-        response.statusText,
-        errorDetails,
-      );
+    err.stack = e.stack;
+  }
+  throw err;
+}
+
+export async function handleResponseNotOk(
```
Comment (DellaBitta) [inline packages/main/src/requests/request.ts:189]: And this?

File: packages/main/src/server/cache-manager.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  CachedContentCreateParams,
+  CachedContentUpdateParams,
+  ListCacheResponse,
+  ListParams,
+} from "../../types/server";
+import { RpcTask } from "./constants";
+import { GoogleGenerativeAIError } from "../errors";
+
+/**
+ * Class for managing GoogleAI file uploads.
```
Comment (DellaBitta) [inline packages/main/src/server/cache-manager.ts:36]: ContentCaches, not file uploads.

File: packages/main/src/server/cache-manager.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  toolConfig?: ToolConfig;
+  systemInstruction?: string | Part | Content;
+
+  constructor(
+    public apiKey: string,
+    private _requestOptions: RequestOptions,
+  ) {}
+
+  /**
+   * Upload a new content cache
```
Comment (DellaBitta) [inline packages/main/src/server/cache-manager.ts:52]: Should be create a new instance of CachedContent, not file upload, and maybe mention that it hits the service..?

File: packages/main/src/server/cache-manager.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+   */
+  async create(
+    createOptions: CachedContentCreateParams,
+  ): Promise<CachedContent> {
+    const newCachedContent: CachedContent = { ...createOptions };
+    if (createOptions.ttlSeconds) {
+      newCachedContent.ttl = createOptions.ttlSeconds.toString() + "s";
+      delete (newCachedContent as CachedContentCreateParams).ttlSeconds;
+    }
+    if (!newCachedContent.model.includes("/")) {
```
Comment (DellaBitta) [inline packages/main/src/server/cache-manager.ts:62]: What if model is empty, shouldn't we report an error? Or do we leave that up to the service?

File: packages/main/src/server/request.ts [hunk 5]
```diff
@@ -65,17 +54,50 @@ export class FilesRequestUrl {
...
+export class FilesRequestUrl extends ServerRequestUrl {
+  constructor(
+    public task: RpcTask,
+    public apiKey: string,
+    public requestOptions?: RequestOptions,
+  ) {
+    super(task, apiKey, requestOptions);
+    const apiVersion = this.requestOptions?.apiVersion || DEFAULT_API_VERSION;
+    const baseUrl = this.requestOptions?.baseUrl || DEFAULT_BASE_URL;
+    let initialUrl = baseUrl;
```
Comment (DellaBitta) [inline packages/main/src/server/request.ts:81]: missing an `l`. `initialUrl`.

File: packages/main/types/function-calling.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
+import { FunctionCallingMode } from "./enums";
```
Comment (DellaBitta) [inline packages/main/types/function-calling.ts:1]: copyright.

File: packages/main/types/server/caching.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+/**
+ * Describes CachedContent interface for sending to the server (if creating)
+ * or received from the server (using getters or list methods).
+ * @public
+ */
+export interface CachedContent extends CachedContentBase {
+  name?: string;
+  ttl?: string;
```
Comment (DellaBitta) [inline packages/main/types/server/caching.ts:38]: Is there a particular format we should define here? Seconds, or something?

File: packages/main/types/server/caching.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  // sent field needs to be protobuf Duration ("3.0001s")
+  ttlSeconds?: number;
+}
+
+/**
+ * Params to pass to {@link GoogleAICacheManager.update}
+ * @public
+ */
+export interface CachedContentUpdateParams {
+  cachedContent: CachedContentCreateParams;
```
Comment (DellaBitta) [inline packages/main/types/server/caching.ts:59]: hm, not a fan of this naming convention but I don't konw what to change it to.

File: packages/main/src/server/request.ts (hunk 5 above)
Comment (hsubox76) [reply packages/main/src/server/request.ts:81]: ? I think I see two Ls?

File: packages/main/types/server/caching.ts (hunk 7 above)
Comment (hsubox76) [reply packages/main/types/server/caching.ts:38]: The input (CachedContentCreateParams) takes `ttlSeconds`. This is the format of the JSON that we send/receive from the API endpoint.

File: packages/main/src/server/request.ts (hunk 5 above)
Comment (DellaBitta) [reply packages/main/src/server/request.ts:81]: Proof positive that I need glasses.

File: packages/main/types/function-calling.ts (hunk 6 above)
Comment (hsubox76) [reply packages/main/types/function-calling.ts:1]: FYI `yarn format` handles this automatically so we don't need to check for it in reviews.

File: packages/main/server/package.json [hunk 9]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "@google/generative-ai-server",
+  "description": "GoogleAI file upload manager",
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply packages/main/server/package.json:3]: Updated!

File: packages/main/src/requests/request.ts [hunk 10]
```diff
@@ -130,89 +130,86 @@ export async function constructRequest(
...
+  }
+
+  if (!response.ok) {
+    await handleResponseNotOk(response, url);
+  }
+
+  return response;
+}
+
+export function handleResponseError(e: Error, url: string): void {
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply packages/main/src/requests/request.ts:173]: Nope, removed export.

File: packages/main/src/requests/request.ts (hunk 1 above)
Comment (hsubox76) [reply packages/main/src/requests/request.ts:189]: Removed export.

File: packages/main/src/server/cache-manager.ts (hunk 2 above)
Comment (hsubox76) [reply packages/main/src/server/cache-manager.ts:42]: Fixed.

File: packages/main/src/server/cache-manager.ts (hunk 3 above)
Comment (hsubox76) [reply packages/main/src/server/cache-manager.ts:58]: I think "upload" is pretty clear that it hits the service? It doesn't create any instance. I notice actually all the variables for headers in this class are called `uploadHeaders` for some reason so I changed them to `headers`. Seems to be a problem with GoogleAIFileManager too but don't want to mess with too many files at once. It's not public facing anyway.

File: packages/main/src/server/cache-manager.ts (hunk 4 above)
Comment (hsubox76) [reply packages/main/src/server/cache-manager.ts:68]: Seems like a good check, added.

File: packages/main/types/server/caching.ts (hunk 7 above)
Comment (DellaBitta) [reply packages/main/types/server/caching.ts:42]: 👍

**Review by DellaBitta — APPROVED** [review]

**Comment by DellaBitta** [comment]
Fixes merged.
Documentation merged.

**Pull request merged**